File: Go/114-flatten-binary-tree-to-linked-list.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func flatten(root *TreeNode)  {
    head := root
    for head != nil {
        //将右子树接在head的前驱节点后面, 并交换左右子树
        if head.Left != nil {
            pre := head.Left
            //找到前驱节点
            for pre.Right != nil {
                pre = pre.Right
            }
            //拼接, 并交换
            pre.Right = head.Right
            head.Left, head.Right = nil, head.Left
        }
        head = head.Right
    }
}

File: Python/114-flatten-binary-tree-to-linked-list.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def flatten(self, root: TreeNode) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        head = root
        while head != None:
            #将右子树接在head的前驱节点后面, 并交换左右子树
            if head.left != None:
                pre = head.left
                #找到前驱节点
                while pre.right != None:
                    pre = pre.right
                #拼接, 并交换
                pre.right = head.right
                head.left, head.right = None, head.left
            head = head.right